controllers/topology/internal/extensions/patches/engine_test.go

```diff
@@ -0,0 +1,541 @@
+							Class:    "default-worker",
+							Name:     "default-worker-topo2",
+							Replicas: pointer.Int32(5),
+						},
+					},
+				},
+			},
+		},
+	}
+
+	// Aggregating Cluster, Templates and ClusterClass into a blueprint.
+	blueprint := &scope.ClusterBlueprint{
+		Topology:                      cluster.Spec.Topology,
+		ClusterClass:                  clusterClass,
+		InfrastructureClusterTemplate: infrastructureClusterTemplate,
+		ControlPlane: &scope.ControlPlaneBlueprint{
+			Template:                      controlPlaneTemplate,
+			InfrastructureMachineTemplate: controlPlaneInfrastructureMachineTemplate,
+		},
+		MachineDeployments: map[string]*scope.MachineDeploymentBlueprint{
+			"default-worker": {
+				InfrastructureMachineTemplate: workerInfrastructureMachineTemplate,
+				BootstrapTemplate:             workerBootstrapTemplate,
+			},
+		},
+	}
+
+	// Create a Cluster using the ClusterClass from above with multiple MachineDeployments
+	// using the same MachineDeployment class.
+	desiredCluster := cluster.DeepCopy()
+
+	infrastructureCluster := builder.InfrastructureCluster(metav1.NamespaceDefault, "infraClusterTemplate1").
+		WithSpecFields(map[string]interface{}{
+			// Add an empty spec field, to make sure the InfrastructureCluster matches
+			// the one calculated by computeInfrastructureCluster.
+			"spec": map[string]interface{}{},
+		}).
+		Build()
+
+	controlPlane := builder.ControlPlane(metav1.NamespaceDefault, "controlPlane1").
+		WithVersion("v1.21.2").
+		WithReplicas(3).
+		// Make sure we're using an independent instance of the template.
+		WithInfrastructureMachineTemplate(controlPlaneInfrastructureMachineTemplate.DeepCopy()).
+		Build()
+
+	desired := &scope.ClusterState{
+		Cluster:               desiredCluster,
+		InfrastructureCluster: infrastructureCluster,
+		ControlPlane: &scope.ControlPlaneState{
+			Object: controlPlane,
+			// Make sure we're using an independent instance of the template.
+			InfrastructureMachineTemplate: controlPlaneInfrastructureMachineTemplate.DeepCopy(),
+		},
+		MachineDeployments: map[string]*scope.MachineDeploymentState{
+			"default-worker-topo1": {
+				Object: builder.MachineDeployment(metav1.NamespaceDefault, "md1").
+					WithVersion("v1.21.2").
+					Build(),
+				// Make sure we're using an independent instance of the template.
+				InfrastructureMachineTemplate: workerInfrastructureMachineTemplate.DeepCopy(),
+				BootstrapTemplate:             workerBootstrapTemplate.DeepCopy(),
+			},
+			"default-worker-topo2": {
+				Object: builder.MachineDeployment(metav1.NamespaceDefault, "md2").
+					WithVersion("v1.20.6").
+					WithReplicas(5).
+					Build(),
+				// Make sure we're using an independent instance of the template.
+				InfrastructureMachineTemplate: workerInfrastructureMachineTemplate.DeepCopy(),
+				BootstrapTemplate:             workerBootstrapTemplate.DeepCopy(),
+			},
+		},
+	}
+	return blueprint, desired
+}
+
+type fakePatchGenerator struct {
```

Comment:
Lets add a comment about what the fakePatchGenerator is (a patch generator that returns the patches provided in input)